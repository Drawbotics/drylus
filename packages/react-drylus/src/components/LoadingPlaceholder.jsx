import React from 'react';
import { css, keyframes } from 'emotion';
import PropTypes from 'prop-types';
import sv from '@drawbotics/drylus-style-vars';


const shimmer = keyframes`
  0% {
    background-position: -1200px 0;
  }
  100% {
    background-position: 1200px 0;
  }
`;

const styles = {
  root: css`
    position: relative;
    pointer-events: none;

    &::after {
      content: ' ';
      position: absolute;
      z-index: 9;
      height: 100%;
      width: 100%;
      top: 0;
      left: 0;
      background: ${sv.neutralLight};
      border-radius: ${sv.defaultBorderRadius};
      overflow: hidden;
      background: linear-gradient(to right,
        ${sv.neutralLight} 8%,
        ${sv.neutralLighter} 18%,
        ${sv.neutralLight} 33%
      );
      background-size: 1200px 100%;
      animation: ${shimmer} 2s forwards infinite linear;
    }
  `,
};


const LoadingPlaceholder = ({ height, width }) => {
  return (
    <div
      className={styles.root}
      style={{ height, width }} />
  );
};


LoadingPlaceholder.propTypes = {
  /** Determines the height of the placeholder */
  height: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]).isRequired,

  /** Determines the height of the placeholder */
  width: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]).isRequired,
};


LoadingPlaceholder.defaultProps = {
  height: sv.defaultMargin,
  width: 200,
};


export default LoadingPlaceholder;