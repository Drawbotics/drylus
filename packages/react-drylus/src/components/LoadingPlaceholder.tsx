import sv from '@drawbotics/drylus-style-vars';
import { css, cx, keyframes } from 'emotion';
import React from 'react';

import { Responsive, Style } from '../types';
import { useResponsiveProps } from '../utils/hooks';

const shimmer = keyframes`
  0% {
    background-position: -1200px 0;
  }
  100% {
    background-position: 1200px 0;
  }
`;

const styles = {
  shimmer: css`
    position: relative;
    pointer-events: none;

    &::after {
      content: ' ';
      position: absolute;
      height: 100%;
      width: 100%;
      top: 0;
      left: 0;
      background: ${sv.neutral};
      border-radius: ${sv.defaultBorderRadius};
      overflow: hidden;
      background-image: linear-gradient(
        to right,
        ${sv.neutral} 8%,
        ${sv.neutralLight} 18%,
        ${sv.neutral} 33%
      );
      background-size: 1200px 100%;
      animation: ${shimmer} 2s forwards infinite linear;
    }
  `,
};

export const placeholderStyles = styles;

export interface LoadingPlaceholderProps {
  /**
   * Determines the height of the placeholder
   * @default sv.defaultMargin
   */
  height?: number | string;

  /**
   * Determines the height of the placeholder
   * @default 200
   */
  width?: number | string;

  /** Reponsive prop overrides */
  responsive?: Responsive<this>;

  /** Used for style overrides */
  style?: Style;

  /** Used for style overrides */
  className?: string;
}

export const LoadingPlaceholder = ({ responsive, ...rest }: LoadingPlaceholderProps) => {
  const { height = sv.defaultMargin, width = 200, style, className } = useResponsiveProps<
    LoadingPlaceholderProps
  >(rest, responsive);

  return <div className={cx(styles.shimmer, className)} style={{ height, width, ...style }} />;
};
