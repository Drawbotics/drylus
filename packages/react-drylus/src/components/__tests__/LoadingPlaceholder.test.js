import React from 'react';
import { create } from 'react-test-renderer';

import LoadingPlaceholder from '../LoadingPlaceholder';


describe('LoadingPlaceholder', () => {
  describe('matches snapshot when', () => {
    it('has no specified sizes', () => {
      const tree = create(
        <LoadingPlaceholder />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has custom sizes', () => {
      const tree = create(
        <LoadingPlaceholder height={200} width={100} />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});