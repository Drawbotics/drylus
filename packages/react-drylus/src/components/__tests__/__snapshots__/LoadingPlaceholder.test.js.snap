// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingPlaceholder matches snapshot when has custom sizes 1`] = `
@keyframes animation-0 {
  0% {
    background-position: -1200px 0;
  }

  100% {
    background-position: 1200px 0;
  }
}

.css-1fxivg0-LoadingPlaceholder__shimmer {
  position: relative;
  pointer-events: none;
}

.css-1fxivg0-LoadingPlaceholder__shimmer::after {
  content: ' ';
  position: absolute;
  z-index: 9;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  background: #d3dce6;
  border-radius: 5px;
  overflow: hidden;
  background: linear-gradient(to right,#d3dce6 8%,#eaeff4 18%,#d3dce6 33% );
  background-size: 1200px 100%;
  -webkit-animation: animation-0 2s forwards infinite linear;
  animation: animation-0 2s forwards infinite linear;
}

<div
  className="css-1fxivg0-LoadingPlaceholder__shimmer"
  style={
    Object {
      "height": 200,
      "width": 100,
    }
  }
/>
`;

exports[`LoadingPlaceholder matches snapshot when has no specified sizes 1`] = `
@keyframes animation-0 {
  0% {
    background-position: -1200px 0;
  }

  100% {
    background-position: 1200px 0;
  }
}

.css-1fxivg0-LoadingPlaceholder__shimmer {
  position: relative;
  pointer-events: none;
}

.css-1fxivg0-LoadingPlaceholder__shimmer::after {
  content: ' ';
  position: absolute;
  z-index: 9;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  background: #d3dce6;
  border-radius: 5px;
  overflow: hidden;
  background: linear-gradient(to right,#d3dce6 8%,#eaeff4 18%,#d3dce6 33% );
  background-size: 1200px 100%;
  -webkit-animation: animation-0 2s forwards infinite linear;
  animation: animation-0 2s forwards infinite linear;
}

<div
  className="css-1fxivg0-LoadingPlaceholder__shimmer"
  style={
    Object {
      "height": "24px",
      "width": 200,
    }
  }
/>
`;
